src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -121,47 +139,105 @@ private void ParseMarkupElement(in SyntaxListBuilder<RazorSyntaxNode> builder)
             if (!NextIs(SyntaxKind.ForwardSlash))
             {
                 // Parsing a start tag
+                var tagStart = CurrentStart;
                 var startTag = ParseStartTag(out var tagName, out var tagMode);
-                builder.Add(startTag);
-
                 if (tagMode == MarkupTagMode.Script)
                 {
-                    ParseJavascriptAndEndScriptTag(builder);
+                    ParseJavascriptAndEndScriptTag(builder, startTag);
+                    return;
+                }
+                else if (tagMode == MarkupTagMode.SelfClosing || tagMode == MarkupTagMode.Invalid)
+                {
+                    // For cases like <foo /> or invalid cases like |<|<p>
+                    var element = SyntaxFactory.MarkupElement(startTag, SyntaxList<RazorSyntaxNode>.Empty, null);
+                    builder.Add(element);
+                    return;
+                }
+                else
+                {
+                    // This is a normal start tag. We need to keep track of it.
+                    var tracker = new TagTracker(tagName, startTag, tagStart, builder);
```

Comment:
Why do we need to keep track of the `PreviousNodes` at all? Can't we just leave them in the builder? If we can't, i'd pass in the elements instead of the builder so it's super clear what the tag tracker is doing with the builder.

Reply:
We need to keep track of previous nodes because otherwise we won't be able to tell the difference between what was already in the builder and what are the children of the current tag. I'll pass in the list instead of the builder